### PR TITLE
Add types argument to __torch_function__

### PR DESCRIPTION
**Pull request by hameerabbasi**
This PR adds the `types` argument to `__torch_function__` as per RFC 0001: https://github.com/pytorch/rfcs/pull/3

### CODE REVIEW

**Comment by ngoldbaum** [comment]
The docs for `__torch_function__` will also need to be updated, see `pytorch/docs/source/notes/extending.rst`.

Ping @cpuhrsch, who may have some insight on how annoying the break in the API for `__torch_function__` will be, at least for `NestedTensor`.

I'm not sure if this would introduce runtime cost but it may be worth either having effectively a try/except in C++ as well as in the Python `handle_torch_function` implementation to accomodate the old `__torch_function__` API, or to generate a nicer error message, in case there are any comsumers of this functionality already with PyTorch 1.4.

File: torch/_overrides.py
Comment (ngoldbaum) [inline torch/_overrides.py]: I think `_get_overloaded_args` guarantees that the entries in `overloaded_args` all have unique types, but it may be worth adding a comment to that effect or doing `tuple(set(map(type, overloaded_args)))` instead.

**Comment by hameerabbasi** [comment]
> The docs for `__torch_function__` will also need to be updated, see `pytorch/docs/source/notes/extending.rst`.

Done.

File: docs/source/notes/extending.rst
Comment (rgommers) [inline docs/source/notes/extending.rst]: "four arguments" now

**Comment by rgommers** [comment]
@hameerabbasi adding benchmark results for this similar to how @ngoldbaum did this before would be good to do next, seems critical. There's docs on using py-spy for this

**Comment by ezyang** [comment]
Yes, benchmarks please!

**Comment by hameerabbasi** [comment]
### Benchmarks
This branch:
```
$ python -m pt.add_test --test_name add_M1_N1_K1_cpu --omp_num_threads 1 --mkl_num_threads 1
# ----------------------------------------
# PyTorch/Caffe2 Operator Micro-benchmarks
# ----------------------------------------
# Tag : None

# Benchmarking PyTorch: add
# Mode: Eager
# Name: add_M1_N1_K1_cpu
# Input: M: 1, N: 1, K: 1, device: cpu
Forward Execution Time (us) : 2.036
```

The commit before it diverges from `master`:
```
$ python -m pt.add_test --test_name add_M1_N1_K1_cpu --omp_num_threads 1 --mkl_num_threads 1
# ----------------------------------------
# PyTorch/Caffe2 Operator Micro-benchmarks
# ----------------------------------------
# Tag : None

# Benchmarking PyTorch: add
# Mode: Eager
# Name: add_M1_N1_K1_cpu
# Input: M: 1, N: 1, K: 1, device: cpu
Forward Execution Time (us) : 2.176
```

Although from running the benchmark a couple of times, everything seems to be well within measurement noise.

**Comment by rgommers** [comment]
@hameerabbasi we went through this a bunch of times before, `asv` is pretty bad at measuring anything with the required accuracy here. Please talk to @ngoldbaum for how to best do this, and/or look at his `__torch_function__` PRs.

**Comment by ngoldbaum** [comment]
@rgommers those *are* the benchmarks I used in those PRs.

**Comment by hameerabbasi** [comment]
@rgommers This is PyTorch's built-in test suite, not ASV. Also, I highly doubt adding an extra argument is going to affect performance in a measurable way. 😉

**Comment by rgommers** [comment]
Hmm, I found the py-spy story much more convincing. The report here is a _gain_ of 140 ns. Given that that's close to 10% of the total call overhead, I think it's important to not wave away?

**Comment by hameerabbasi** [comment]
[Here are the two `py-spy` SVGs, one for each case.](https://github.com/pytorch/pytorch/files/4308509/py-spy-perf.zip)

Like I said, this is due to noise. I could re-run tests, but I ran each test a couple of times but _reported_ the first run of each. The subsequent runs show a large variation, larger than is reported here (+-200 ns IIRC).

So, it's possible due to pure _noise_ that my addition had a reported performance better than `master`, even if that isn't based in reality.

**Comment by hameerabbasi** [comment]
Also, we're looking at the case of a one-element add, where the computation is basically nothing, and the whole cost is the machinery.

**Comment by ezyang** [comment]
The diff looks basically fine but I'd love to see the benchmarks

**Comment by ezyang** [comment]
@hameerabbasi Some questions about benchmark hygiene:
* What compile options did you use to compile PyTorch?
* Can you test on functions with more arguments? We expect the cost of allocating the types python tuple to scale linearly with the number of tensor-like arguments. If it remains invisible even for a function with a lot of tensor arguments, there's something wrong with the benchmark.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (hameerabbasi) [inline torch/csrc/utils/python_arg_parser.cpp]: I pre-allocated the `std::vector` here so that a lower amount of allocations happen.

**Comment by hameerabbasi** [comment]
> Can you test on functions with more arguments?

Sure, but this won't have much effect, read on.

> We expect the cost of allocating the types python tuple to scale linearly with the number of tensor-like arguments. If it remains invisible even for a function with a lot of tensor arguments, there's something wrong with the benchmark.

This statement is false. `overloaded_args` contains one argument of each type, and so we'd need to have a type factory spawning a large number of types for this to scale, I'll do that but it's clearly not a very practical use-case.

**Comment by hameerabbasi** [comment]
As per discussion with @rgommers,

Here are benchmarks for a `Tensor`, a subclass of `Tensor`, and a third case with `__torch_function__` implemented and a realistic implementation forwarding to `_make_subclass` for add.

This is for a million repetitions, so mentally change seconds to microseconds. Again, noise is high, but the overhead seems pretty small in all cases, and is theoretically zero for the first two cases.

Base:
```
(pytorch-cuda-dev) habbasi@qgpu2:~$ time python bench-tensor-only.py

real	0m2.293s
user	0m2.225s
sys	0m0.042s
(pytorch-cuda-dev) habbasi@qgpu2:~$ time python bench-tensorlike.py

real	0m2.964s
user	0m2.899s
sys	0m0.041s
(pytorch-cuda-dev) habbasi@qgpu2:~$ time python bench-withtorchfunction.py

real	0m4.770s
user	0m4.667s
sys	0m0.077s
```

Branch:
```
(pytorch-cuda-dev) habbasi@qgpu2:~$ time python bench-tensor-only.py

real	0m2.209s
user	0m2.120s
sys	0m0.062s
(pytorch-cuda-dev) habbasi@qgpu2:~$ time python bench-tensorlike.py

real	0m2.883s
user	0m2.806s
sys	0m0.052s
(pytorch-cuda-dev) habbasi@qgpu2:~$ time python bench-withtorchfunction.py

real	0m5.067s
user	0m4.954s
sys	0m0.085s
```

Screenshot of `py-spy` run on the only theoretically affected case

**Comment by rgommers** [comment]
Thanks @hameerabbasi that looks good!

@ezyang I think we'll want to put this overhead benchmarking into a procedure/script, since it's so hard to get the right O(100 ns) numbers visible. Would a new script with procedure in its docstring, and linking to that from `torch/_overrides.py`, be the right place for that?

@hameerabbasi not sure that's the right flamegraph, since the overhead in `withtorchfunction` should be almost double of the other ones hence some `__torch_function__`-related calls should be clearly identifiable?

**Comment by hameerabbasi** [comment]
> not sure that's the right flamegraph
You're right, I updated it.

**Comment by ezyang** [comment]
> Would a new script with procedure in its docstring, and linking to that from torch/_overrides.py, be the right place for that?

That sounds reasonable to me. It will be difficult to get people to find out about the script but if we at least know about it ourselves that will be good enough for now.

**Comment by rgommers** [comment]
> It will be difficult to get people to find out about the script

We can just point to it from the critical parts of the `__torch_function__` mechanism. One only needs these benchmarks when touching those parts, so it shouldn't be too bad.

**Comment by hameerabbasi** [comment]
Should I leave that to this PR or a follow-up?

**Comment by rgommers** [comment]
Separate PR seems better; that PR could be merged straight away, while I think this one we may want to consider together with the `make_subclass` one and the to-be-submitted methods overrides one.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
can you please merge with master

**Comment by cpuhrsch** [comment]
Just FYI https://github.com/pytorch/pytorch/pull/34240 is still in progress of landing.

**Comment by ezyang** [comment]
Thanks Christian. We'll hold this PR for that one, and then remerge.

**Comment by hameerabbasi** [comment]
Benchmark PR at #34645.

**Comment by cpuhrsch** [comment]
#34240 was merged

**Comment by hameerabbasi** [comment]
Re-merged.

**Comment by hameerabbasi** [comment]
Looks like internal tests are failing.

**Comment by ezyang** [comment]
@hameerabbasi I can't seem to push to your branch (did you fail to allow maintainers to update the branch). Could you please merge with master?

**Comment by hameerabbasi** [comment]
> did you fail to allow maintainers to update the branch

That might've been unchecked by default, and I can't find the checkbox where it usually is on the side of the PR.

**Comment by ezyang** [comment]
@cpuhrsch heads up this will break you

**Pull request closed without merging**
